### ISSUE TITLE
Question about beam_sample: using two softmax?

### ISSUE THREAD

**Issue by binshengliu** [huggingface/transformers#8702]
## Environment info

     
- `transformers` version: 3.5.1
- Platform: Linux-4.15.0-122-generic-x86_64-with-glibc2.10
- Python version: 3.8.5
- PyTorch version (GPU?): 1.7.0+cu101 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes

### Who can help
@patrickvonplaten @TevenLeScao

## Information

Model I am using (Bert, XLNet ...): Bart

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [X] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [X] my own task or dataset: (give details below)

## To reproduce

I just have some confusions about the current beam_sample code. Please disregard if I misunderstood.

I noticed in current beam_sample code there are two softmax operations to produce the token probabilities, line 1161 and 1171 in the following snippet. After 1171, the final probabilities would be similar to `F.softmax(F.log_softmax(next_token_logits, dim=-1), dim=-1)` which is very different from what we usually get by `softmax(logits)`. Similarly in [top-p filtering](https://github.com/huggingface/transformers/blob/9c0afdaf7b091c341072b432ad6ee17ba7a5016b/src/transformers/generation_logits_process.py#L184) `F.softmax` is used on log values. But shouldn't we use `exp` to recover the probabilities in these cases?

https://github.com/huggingface/transformers/blob/9c0afdaf7b091c341072b432ad6ee17ba7a5016b/src/transformers/generation_utils.py#L1161-L1171

**Comment by binshengliu** [comment]
After submitting I realized `F.softmax(F.log_softmax())` is equivalent to `F.softmax()`. The different probability values I noticed comes from the normalization of multiple beams in one dimension (line 1169). That doesn't change the behaviour of the sampling.
